Added more tests to Form to increase coverage

jcfilben opened the pull request:
#### What does this PR do?
Adds additional testing to Form to increase testing coverage.
Testing Coverage for Form.js:
| Type | Before | After |
| ----- | ------ | ------ |
| Statements | 92.91% | 100% |
| Branches | 87.21% | 96.51% |
| Functions | 86.21% | 100% |
| Lines | 95.76% | 100% |

Total Coverage:
| Type | Before | After |
| ----- | ------ | ------ |
| Statements | 82.97% | 83.42% |
| Branches | 73.24% | 73.84% |
| Functions | 80.22% | 80.76% |
| Lines | 84.07% | 84.43% |

#### Additional Notes:
Line 158 is covered by SelectForm-test.js and brings the coverage up to 100% for Form.

#### Tests added/changed and explanation:

'validate on blur' - changed test to add focus events. When the active element in the document changes the previously focused element gets blurred. So to trigger a blur event you first have to have a focus event. This adds coverage to lines 224-235.

'uncontrolled without name' - a test for an uncontrolled form that doesn't have the name prop. Adds else branch coverage to line 142.

'controlled reset' - a test for a controlled form when reset is called and value is set. Adds else branch coverage to line 246.

'controlled reset without value' - a test for a controlled form when reset is called and value is not set and onChange is called. Adds coverage to line 248 and else branch coverage for line 254.

#### Where should the reviewer start?
Form-test.js

#### What testing has been done on this PR?
`yarn test` and coverage report

#### How should this be manually tested?

#### Any background context you want to provide?

#### What are the relevant issues?
closes #4227

#### Screenshots (if appropriate)
<img width="1772" alt="Screen Shot 2020-06-24 at 4 01 50 PM" src="https://user-images.githubusercontent.com/54560994/85632153-06fcb200-b634-11ea-85c3-35c91c29bcfb.png">

<img width="1774" alt="Screen Shot 2020-06-24 at 3 59 16 PM" src="https://user-images.githubusercontent.com/54560994/85631956-ac635600-b633-11ea-8a7e-f5fd8f4a5396.png">


#### Do the grommet docs need to be updated?
No

#### Should this PR be mentioned in the release notes?
No

#### Is this change backwards compatible or is it a breaking change?
Backwards compatible

JoeButler7 commented on src/js/components/Form/__tests__/Form-test.js:
```diff
@@ -581,4 +587,80 @@ describe('Form', () => {
+      );
+    };
+    const { getByPlaceholderText, getByText, queryByText } = render(
+      <Grommet>
```
Same as above, also it seems a bit redundant to set the component to a variable in the test and then render it, why not just put all of the code in `render` call because it is in the test anyways?

jcfilben replied:
Updated so the code is in render instead of test variable. There are certain tests that do need to be set up like this to work but this was not one of them. If the test has hooks (ex. `const [value, setValue] = React.useState({ test: '' });`
 and `const onChange = React.useCallback(nextValue => setValue(nextValue), []);`) then the code has to be outside of render.

JoeButler7 commented on src/js/components/Form/__tests__/Form-test.js:
```diff
@@ -581,4 +587,80 @@ describe('Form', () => {
+      );
+    };
+    const { getByPlaceholderText, getByText, queryByText } = render(
+      <Grommet>
```
Here you are wrapping `<Test/>` inside of a `<Grommet/>` when it `<Test/>` is already wrapped inside of a `<Grommet/>` component

jcfilben replied:
Updated so it is no longer double wrapped in <Grommet>

JoeButler7 commented on src/js/components/Form/__tests__/Form-test-controlled.js:
```diff
@@ -222,4 +225,61 @@ describe('Form controlled', () => {
+    expect(queryByText('Input has changed')).toBeNull();
+  });
+
+  test('form with select', async () => {
```
Why is this test async?

jcfilben replied:
Changed test so it is no longer async

JoeButler7 commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -277,8 +363,8 @@ describe('Form uncontrolled', () => {
       <Grommet>
-        {/* this test continues running forever if the whole event 
-                passed to onSubmit */}
+        {/* this test continues running forever if the whole event passed to 
```
Confused as to why this is here

ShimiSun replied:
Can you update on the resolution?

jcfilben replied:
This comment was originally here and I changed the spacing which is why it is marked as being changed, to resolve it I just changed the spacing back to the original

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -1,14 +1,100 @@
-import { cleanup, render, fireEvent } from '@testing-library/react';
+
 import 'jest-styled-components';
+import 'jest-axe/extend-expect';
+import 'regenerator-runtime/runtime';
 
+import { cleanup, render, fireEvent } from '@testing-library/react';
+import { axe } from 'jest-axe';
```
Let's try to organize the packages with a section per the external libraries and a section for Grommet imports.

jcfilben replied:
Separated libraries by Grommet and external

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -1,14 +1,100 @@
+describe('Form accessibility', () => {
+  afterEach(cleanup);
+
+  test('should have no accessibility violations', async () => {
```
```suggestion
  test('TextInput in Form should have no accessibility violations', async () => {
```

jcfilben replied:
Changed accessibility test names

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -1,14 +1,100 @@
+    expect(results).toHaveNoViolations();
+  });
+
+  test('form with select accessibility', async () => {
```
```suggestion
  test('Select in Form should have no accessibility violations', async () => {
```

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -1,14 +1,100 @@
+    expect(results).toHaveNoViolations();
+  });
+
+  test('form with checkbox accessiility', async () => {
```
```suggestion
  test('CheckBox in Form should have no accessibility violations', async () => {
```

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -1,14 +1,100 @@
+    expect(results).toHaveNoViolations();
+  });
+
+  test('form with textinput accessiility', async () => {
```
```suggestion
  test('FormField with an explicit TextInput child should have no accessibility violations', async () => {
```
I would move this axe test to be right after the first test of TextInput accessibility.

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -1,14 +1,100 @@
+    expect(results).toHaveNoViolations();
+  });
+
+  test('box with textinput accessibility', async () => {
```
```suggestion
  test('Box with TextInput in Form should have no accessibility violations', async () => {
```

jcfilben commented:
### Update:

- Added `a11yTitle` to tests other than the accessibility tests (they already had `a11yTitle`).
- Removed TextInput from Select test (line 272 in `Form-test-controlled.js`).
- Added a line between external libraries and Grommet imports in `Form-test-uncontrolled.js`.
- Changed names of accessibility tests (in some cases the names made the line length too long and I had to shorten the test name and put a comment above the test with the longer test name. Let me know if there is a different/better way to do this).

jcfilben commented on src/js/components/Form/__tests__/Form-test-controlled.js:
```diff
@@ -222,4 +224,60 @@ describe('Form controlled', () => {
+        <Grommet>
+          <Form>
+            <FormField>
+              <Select
```
Added a11yTitle

jcfilben commented on src/js/components/Form/__tests__/Form-test-controlled.js:
```diff
@@ -222,4 +224,60 @@ describe('Form controlled', () => {
+                onChange={({ option }) => setValue(option)}
+              />
+            </FormField>
+            <FormField name="test" required placeholder="test input 2" />
```
Removed TextInput line

JoeButler7 commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -1,14 +1,105 @@
+  afterEach(cleanup);
+
+  // TextInput in Form should have no accessibility violations
+  test('TextInput-Form should have no accessibility violations', async () => {
```
you can use forward ticks `` to create multiline strings here

ShimiSun commented on src/js/components/Form/__tests__/Form-test-controlled.js:
```diff
@@ -222,4 +224,65 @@ describe('Form controlled', () => {
+      const [value, setValue] = React.useState('medium');
+      return (
+        <Grommet>
+          <Form>
```
Since Form isn't holding either `value` or onChange, should it be on the controlled file test?

jcfilben replied:
Good point this test should be in the uncontrolled test file, I will go back and change it

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -302,75 +392,212 @@ describe('Form uncontrolled', () => {
+    fireEvent.change(getByPlaceholderText('test input'), {
+      target: { value: 'v' },
+    });
+    expect(container.firstChild).toMatchSnapshot();
```
Can we add an assertion on the fetched input instead of taking the snapshot, unless you are positive the snapshot is capturing the new value.

jcfilben replied:
Removed snapshot and added assertion

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -302,75 +392,212 @@ describe('Form uncontrolled', () => {
+    expect(queryByText('Input has changed')).toBeNull();
+  });
+
+  test('disabled', () => {
```
```suggestion
  test('disabled FormField', () => {
```

jcfilben replied:
Changed name from 'disabled' to 'disabled FormField'

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -302,75 +392,212 @@ describe('Form uncontrolled', () => {
+        </Form>
+      </Grommet>,
+    );
+    fireEvent.change(getByPlaceholderText('test input'), {
+      target: { value: 'v' },
+    });
```
Similar comment regarding assertion vs snapshot.

jcfilben replied:
Removed snapshot and added assertion

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -302,75 +392,212 @@ describe('Form uncontrolled', () => {
+  });
+
+  test('custom component', () => {
+    const TextIn = ({ name, value, onChange }) => {
```
```suggestion
    const CustomTextInput = ({ name, value, onChange }) => {
```

jcfilben replied:
Changed name from 'TextIn' to 'CustomTextInput'

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -302,75 +392,212 @@ describe('Form uncontrolled', () => {
+      <Grommet>
+        <Form onChange={onChange}>
+          <FormField required placeholder="test input">
+            <TextIn value="Hello World" name="test" onChange={onChange} />
```
```suggestion
            <CustomTextInput value="Hello World" name="test" onChange={onChange} />
```

jcfilben replied:
Changed name from 'TextIn' to 'CustomTextInput'

ShimiSun commented on src/js/components/Form/__tests__/Form-test-uncontrolled.js:
```diff
@@ -573,4 +800,31 @@ describe('Form uncontrolled', () => {
+      );
+    };
+    const { getByPlaceholderText, container } = render(<Test />);
+    fireEvent.change(getByPlaceholderText('test input'), {
+      target: { value: 'small' },
+    });
```
Similar comment regarding missing assertion.

jcfilben replied:
Added assertion

jcfilben commented on src/js/components/Form/__tests__/Form-test-controlled.js:
```diff
@@ -222,4 +223,38 @@ describe('Form controlled', () => {
+      );
+    };
+    const { getByPlaceholderText, getByText, queryByText } = render(<Test />);
+    fireEvent.change(getByPlaceholderText('test input'), {
+      target: { value: 'Input has changed' },
+    });
```
Added assertion

jcfilben commented:
Form the form test with select I changed it to have an assertion statement instead of a snapshot but I wasn't sure if the test should be included under Form testing in the first place because Joe has similar tests under Select-Form testing

ShimiSun commented:
@jcfilben I would like to have a test of Select within Form. It's only an assertion test so it should be fine. we can ask Joe to have a different flavor of the test on Select.

ShimiSun commented:
Great work @jcfilben !!
